File: datasets/Part 3 - Classification/Section 19 - Decision Tree Classification/decision_tree_classification_sbn.R
# ARBOLES DE DECISION
# Importar dataset
dataset <- read.csv('Social_Network_Ads.csv')
dataset = dataset[, 3:5]

# Codificar la variable de clasificacion como factor
dataset$Purchased = factor(dataset$Purchased,
                           levels = c(0, 1))

# Dividir el dataset en entrenamiento y prueba
#install.packages('caTools')
library(caTools)
set.seed(123) #SOLO ES PARA QUE SIEMPRE SELECCIONE LOS MISMOS ELEMENTOS
split <- sample.split(dataset$Purchased, SplitRatio = 0.75)
trainning_set <- subset(dataset, split == TRUE)
testing_set <- subset(dataset, split == FALSE)

# Ajustar la clasificacion con el conjunto de entrenamiento
library(rpart)
classifier <- rpart(formula = Purchased ~ .,
                   data = trainning_set)

# Prediccion de la clasificacion con el conjunto de prueba
y_pred <- predict(classifier,
                  newdata = testing_set[, -3],
                  type = 'class')

# Evaluar resultados con matriz de confusion
conf_matrix <- table(testing_set[, 3], y_pred)

# Visualizacion de los resultados
#install.packages('ElemStatLearn')
library(ElemStatLearn)
#set <- trainning_set
set <- testing_set
X1 <- seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.05)
X2 <- seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 250)
grid_set <- expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid <- predict(classifier,
                  newdata = grid_set,
                  type = 'class')
plot(set[, -3],
     main = 'Clasificador (Conjunto de Entrenamiento)',
     xlab = 'Edad',
     ylab = 'Sueldo Estimado',
     xlim = range(X1),
     ylim = range(X2))
contour(X1,
        X2,
        matrix(as.numeric(y_grid),
               length(X1),
               length(X2)),
        add = TRUE)
points(grid_set,
       pch = '.',
       col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set,
       pch = 21,
       bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

# VISUALIZACION DE ARBOL DE DECISION
plot(classifier)
text(classifier)

File: datasets/Part 3 - Classification/Section 19 - Decision Tree Classification/decision_tree_classification_sbn.py
# -*- coding: utf-8 -*-
"""
Created on Fri Feb  7 09:41:01 2020

@author: jguerrero
"""

# ARBOLES DE DECISION
# Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('Social_Network_ads.csv')

X = dataset.iloc[:, 2:4].values
y = dataset.iloc[:, 4].values

# Dividir el dataset en entrenamiento y prueba
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0) # random_state = 0 SOLO ES PARA QUE SIEMPRE SELECCIONE LOS MISMOS ELEMENTOS

# Ajustar la clasificacion con el conjunto de entrenamiento
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Prediccion de la clasificacion con el conjunto de prueba
y_pred = classifier.predict(X_test)

# Evaluar resultados con matriz de confusion
from sklearn.metrics import confusion_matrix
conf_matrix = confusion_matrix(y_test, y_pred)

# Visualizacion de los resultados
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 1),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 500))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green'))(i), label = j)
plt.title('Arbol de decision (Conjunto de Entrenamiento)')
plt.xlabel('Edad')
plt.ylabel('Sueldo Estimado')
plt.legend()
plt.show()

from sklearn.tree import plot_tree
plot_tree(classifier)